tests/resolve_imports/file_parser.cpp

```diff
@@ -0,0 +1,124 @@
+limitations under the License.
+*/
+
+#include "gtest/gtest.h"
+
+#include <algorithm>
+#include <fstream>
+#include <iostream>
+#include <libcellml>
+#include <sstream>
+#include <vector>
+
+// generated with test resource locations
+#include "test_resources.h"
+
+TEST(ResolveImports, resolveSineModelFromFile) {
+    std::ifstream t(TestResources::getLocation(
+                    TestResources::CELLML_SINE_MODEL_RESOURCE));
+
+    std::stringstream buffer;
+    buffer << t.rdbuf();
+
+    libcellml::Parser p;
+    libcellml::ModelPtr model = p.parseModel(buffer.str());
+
+    EXPECT_EQ(0u, p.errorCount());
+    EXPECT_FALSE(model->hasUnresolvedImports());
+}
+
+TEST(ResolveImports, resolveSineImportsModelFromFile) {
+    std::string sineModelLocation = TestResources::getLocation(
+                TestResources::CELLML_SINE_IMPORTS_MODEL_RESOURCE);
+    std::ifstream t(sineModelLocation);
+    std::stringstream buffer;
+    buffer << t.rdbuf();
+
+    libcellml::Parser p;
+    libcellml::ModelPtr model = p.parseModel(buffer.str());
+    EXPECT_EQ(0u, p.errorCount());
+
+    EXPECT_TRUE(model->hasUnresolvedImports());
+    model->resolveImports(sineModelLocation);
+    EXPECT_FALSE(model->hasUnresolvedImports());
+}
+
+TEST(ResolveImports, resolveComplexImportsModelFromFile) {
+    std::string modelLocation = TestResources::getLocation(
+                TestResources::CELLML_COMPLEX_IMPORTS_MODEL_RESOURCE);
+    std::ifstream t(modelLocation);
+    std::stringstream buffer;
+    buffer << t.rdbuf();
+
+    libcellml::Parser p;
+    libcellml::ModelPtr model = p.parseModel(buffer.str());
+    EXPECT_EQ(0u, p.errorCount());
+
+    EXPECT_TRUE(model->hasUnresolvedImports());
+    model->resolveImports(modelLocation);
+    EXPECT_FALSE(model->hasUnresolvedImports());
+}
+
+TEST(ResolveImports, resolveUnitsImportFromFile) {
+    std::string modelLocation = TestResources::getLocation(
+        TestResources::CELLML_UNITS_IMPORT_MODEL_RESOURCE);
+    std::ifstream t(modelLocation);
+    std::stringstream buffer;
+    buffer << t.rdbuf();
+
+    libcellml::Parser p;
+    libcellml::ModelPtr model = p.parseModel(buffer.str());
+
+    EXPECT_EQ(0u, p.errorCount());
+
+    EXPECT_TRUE(model->hasUnresolvedImports());
+    model->resolveImports(modelLocation);
+    EXPECT_FALSE(model->hasUnresolvedImports());
+}
+
+TEST(ResolveImports, resolveImportsFromFileLevel0) {
+    std::string modelLocation = TestResources::getLocation(
+        TestResources::CELLML_IMPORT_LEVEL0_MODEL_RESOURCE);
+    std::ifstream t(modelLocation);
+    std::stringstream buffer;
+    buffer << t.rdbuf();
+
+    libcellml::Parser p;
+    libcellml::ModelPtr model = p.parseModel(buffer.str());
+
+    EXPECT_EQ(0u, p.errorCount());
+
+    EXPECT_TRUE(model->hasUnresolvedImports());
+    model->resolveImports(modelLocation);
+    EXPECT_FALSE(model->hasUnresolvedImports());
+}
+
+TEST(ResolveImports, resolveImportsFromFileLevel0Unresolvable) {
```

Comment:
It would be nice if `libCellML` was to provide some information about why imports couldn't be resolved.

Reply:
It would, but I think that is the job here. We need to extend the validator to do more thorough validation through the import hierarchy on demand. I think that is the way to go, keep this functionality just for resolving the imports that can be resolved.

Reply:
Ok, agreed.